File: Analyze Data with R/Page Visits Funnel.R
---
  title: "Page Visits Funnel"
output: html_notebook
---
# load packages
library(readr)
library(dplyr)
```

# load data
visits <- read_csv("visits.csv")
cart <- read_csv("cart.csv")
checkout <- read_csv("checkout.csv")
purchase <- read_csv("purchase.csv")
```

# inspect data frames
head(visits)
head(cart)
head(checkout)
head(purchase)
```

# define visits_cart here:
visits_cart <- visits %>% 
  left_join(cart)


# How long is the visits data frame?:
total_visits <- nrow(visits)

# How many of the timestamps are NA for the column cart_time:
visit_no_cart <- cart %>%
  filter(is.na(cart_time))

visit_no_cart
visit_no_cart_count <- nrow(visit_no_cart)
visit_no_cart_count

# calculate visit_no_cart_percent here:
visit_no_cart_percent <- visit_no_cart_count/total_visits
visit_no_cart_percent

# define cart_checkout here:
cart_checkout <- cart %>% 
  left_join(checkout)

# define total_carts here:
total_carts = nrow(cart)
total_carts

# define cart_no_checkout here:
cart_no_checkout <- cart_checkout %>% 
  filter(is.na(checkout_time))
cart_no_checkout

cart_no_checkout_count <- nrow(cart_no_checkout)
cart_no_checkout_count

# calculate cart_no_checkout_percent here:
cart_no_checkout_percent <- cart_no_checkout_count/total_carts
cart_no_checkout_percent

# define all_data here:
all_data <- visits_cart %>%
  left_join(cart_checkout)

head(all_data)

# define total_checkout here:
total_checkout <- nrow(checkout)
total_checkout

# define checkout_no_purchase here:
checkout_no_purchase <- purchase %>% 
  filter(is.na(purchase_time))
checkout_no_purchase

# calculate checkout_no_purchase_percent here:
checkout_no_purchase_percent <- checkout_no_purchase / total_checkout
checkout_no_purchase_percent

# update all_data with time_to_purchase column here:
all_data <- all_data %>%
  mutate(time_to_purchase = purchase_time - visit_time)


# inspect the updated all_data data frame here:

head(all_data)

# define time_to_purchase here:
time_to_purchase <- all_data %>%
  summarize(mean_time = mean(time_to_purchase, na.rm = TRUE))
time_to_purchase

File: Data-Science-Path/Pandas/Page Visits Funnel.py
# Multiple tables in Pandas

# Page Visits Funnel

import codecademylib
import pandas as pd

visits = pd.read_csv('visits.csv', parse_dates=[1])
cart = pd.read_csv('cart.csv', parse_dates=[1])
checkout = pd.read_csv('checkout.csv', parse_dates=[1])
purchase = pd.read_csv('purchase.csv', parse_dates=[1])

# Inspect the DataFrames using print and head
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())

# Combine visits and cart using a left merge
visits_cart = pd.merge(visits, cart, how='left')
print(visits_cart)

# How many of the timestamps are null for the column cart_time
null_cart_times = len(visits_cart[visits_cart.cart_time.isnull()])

# What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?
print(float(null_cart_times/len(visits_cart)))

# Repeat the left merge for cart and checkout and count null values. 
# What percentage of users put items in their cart, but did not proceed to checkout?
cart_checkout = pd.merge(cart, checkout, how='left')
print(cart_checkout)

print(float(len(cart_checkout[cart_checkout.checkout_time.isnull()])/len(cart_checkout)))

# Merge all four steps of the funnel, in order, using a series of left merges. Save the results to the variable all_data.
# Examine the result using print and head.
all_data = visits.merge(cart, how="left").merge(checkout, how="left").merge(purchase, how="left")
print(all_data.head())

# What percentage of users proceeded to checkout, but did not purchase a t-shirt?
checkout_purchase = pd.merge(checkout, purchase, how ="left")
checkout_purchase_rows = len(checkout_purchase)
null_purchase_times = len(checkout_purchase[checkout_purchase.purchase_times.isnull()])
print(float(null_purchase_timces/checkout_purchase_rows))

# Using the giant merged DataFrame all_data that you created, let’s calculate the average time from initial visit to final purchase
all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time
print(all_data.time_to_purchase)

# Calculate the average time to purchase 
print(all_data.time_to_purchase.mean())